Clang wrapper for Windows isn't working

These are some of the issues of our Clang wrapper for Windows:
```console
% julia --compile=min -e 'using BinaryBuilderBase; BinaryBuilderBase.runshell(Platform("x86_64", "windows"))'                          
sandbox:${WORKSPACE} # echo 'int main(){return 0;}' | clang -x c -
clang-13: warning: argument unused during compilation: '--gcc-toolchain=/opt/x86_64-w64-mingw32' [-Wunused-command-line-argument]
clang-13: warning: argument unused during compilation: '-stdlib=libstdc++' [-Wunused-command-line-argument]
/opt/x86_64-w64-mingw32/bin/x86_64-w64-mingw32-ld: cannot find -lgcc
/opt/x86_64-w64-mingw32/bin/x86_64-w64-mingw32-ld: cannot find -lgcc_eh
/opt/x86_64-w64-mingw32/bin/x86_64-w64-mingw32-ld: cannot find -lgcc
/opt/x86_64-w64-mingw32/bin/x86_64-w64-mingw32-ld: cannot find -lgcc_eh
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
sandbox:${WORKSPACE} # echo -e '#include <limits>\nint main(){return 0;}' | clang -x c++ -
clang-13: warning: argument unused during compilation: '--gcc-toolchain=/opt/x86_64-w64-mingw32' [-Wunused-command-line-argument]
<stdin>:1:10: fatal error: 'limits' file not found
#include <limits>
         ^~~~~~~~
1 error generated.
```
We actually already do https://github.com/JuliaPackaging/BinaryBuilderBase.jl/blob/4b21cde2105830e287271c323d909b9ee8a64c5e/src/Runner.jl#L318-L319 to make Clang find the GCC toolchain, but it doesn't seem to be effective on Windows, maybe Clang expects a different tree structure, or maybe it's just ignored (why?), given the warning about the argument being unused.

I tried to address this issue in #247, but it's a horrible ad-hoc kludge.  I opened this issue because I feel like we should be able to make this work more nicely than putting a bunch of random `-I...` and `-L...` arguments.

--- a/Artifacts.toml
+++ b/Artifacts.toml
@@ -2770,6 +2770,28 @@ os = "linux"
     sha256 = "baf8f1f42219ba94d0bed5a8d2533bd4bb9f8e109691d930e0da57f2746ac02a"
     url = "https://github.com/JuliaPackaging/Yggdrasil/releases/download/LLVMBootstrap-v13.0.1+1/LLVMBootstrap.v13.0.1.x86_64-linux-musl.unpacked.tar.gz"
 
+[["LLVMBootstrap.v16.0.6.x86_64-linux-musl.squashfs"]]
+arch = "x86_64"
+git-tree-sha1 = "69c66aef8882dafafa179cf7152fcdf825f19591"
+lazy = true
+libc = "musl"
+os = "linux"
+
+    [["LLVMBootstrap.v16.0.6.x86_64-linux-musl.squashfs".download]]
+    sha256 = "e0197e2ed9f96b9e7e5c369068d54a9a85a08588739f6b84f08ad5bde825553e"
+    url = "https://github.com/JuliaPackaging/Yggdrasil/releases/download/LLVMBootstrap-v16.0.6/LLVMBootstrap.v16.0.6.x86_64-linux-musl.squashfs.tar.gz"
+
+[["LLVMBootstrap.v16.0.6.x86_64-linux-musl.unpacked"]]
+arch = "x86_64"
+git-tree-sha1 = "ef7ca988b60e65d8e6dd139bc7c18864a760ffa8"
+lazy = true
+libc = "musl"
+os = "linux"
+
+    [["LLVMBootstrap.v16.0.6.x86_64-linux-musl.unpacked".download]]
+    sha256 = "a4cc72616d41db70d8452b8706078800a15c7e01f20877634029ff36a5083b07"
+    url = "https://github.com/JuliaPackaging/Yggdrasil/releases/download/LLVMBootstrap-v16.0.6/LLVMBootstrap.v16.0.6.x86_64-linux-musl.unpacked.tar.gz"
+
 [["LLVMBootstrap.v6.0.1.x86_64-linux-musl.squashfs"]]
 arch = "x86_64"
 git-tree-sha1 = "474b55ee699607660a7375f4e38afc9a76583c41"

--- a/src/Rootfs.jl
+++ b/src/Rootfs.jl
@@ -804,7 +804,7 @@ function expand_cxxstring_abis(platform::AbstractPlatform; skip=Sys.isbsd)
         p["cxxstring_abi"] = "cxx11" #Clang only seems to generate cxx11 abi
         return [p]
     end
-                                                                                                                                 
+
     # Otherwise, generate new versions!
     map(["cxx03", "cxx11"]) do abi
         p = deepcopy(platform)

--- a/src/Runner.jl
+++ b/src/Runner.jl
@@ -322,6 +322,11 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
             "--sysroot=/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root",
         ])
         if !Sys.isbsd(p)
+            if cxxstring_abi(p) == "cxx11"
+                push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=1")
+            elseif cxxstring_abi(p) == "cxx03"
+                push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=0")
+            end
             if iscxx
                 append!(flags, [
                     # Link with libstdc++ when compiling c++ on non-BSDs

--- a/src/Runner.jl
+++ b/src/Runner.jl
@@ -141,6 +141,15 @@ end
 prefer_clang(p::AbstractPlatform) =
     Sys.isbsd(p) || sanitize(p) in ("memory", "memory_origins", "address")
 
+# Add the string ABI define
+function add_cxx_abi(p::AbstractPlatform, flags::Vector{String})
+    if cxxstring_abi(p) == "cxx11"
+        push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=1")
+    elseif cxxstring_abi(p) == "cxx03"
+        push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=0")
+    end
+end
+
 """
     generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::AbstractString,
                                 host_platform::AbstractPlatform = $(repr(default_host_platform)),
@@ -162,6 +171,7 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
                                      allow_unsafe_flags::Bool = false,
                                      lock_microarchitecture::Bool = true,
                                      bootstrap::Bool = !isempty(bootstrap_list),
+                                     gcc_version::Union{Nothing,VersionNumber}=nothing,
                                      )
     # Wipe that directory out, in case it already had compiler wrappers
     rm(bin_path; recursive=true, force=true)
@@ -322,21 +332,15 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
             "--sysroot=/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root",
         ])
         if !Sys.isbsd(p)
-            if cxxstring_abi(p) == "cxx11"
-                push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=1")
-            elseif cxxstring_abi(p) == "cxx03"
-                push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=0")
-            end
+            add_cxx_abi(p, flags)
             if iscxx
                 append!(flags, [
                     # Link with libstdc++ when compiling c++ on non-BSDs
                     "-stdlib=libstdc++"
             ])
-            push!(flags, "-nobuiltininc")
             end
         end
-        if Sys.islinux(p)
-            gcc_version, llvm_version = select_compiler_versions(p, compilers)
+        if Sys.islinux(p) && !isnothing(gcc_version)
             append!(flags, ["--gcc-install-dir=/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)"])
         end
         if Sys.iswindows(p)
@@ -422,15 +426,12 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
             add_system_includedir(flags)
         end
 
-        if !Sys.isbsd(p)
-            gcc_version, llvm_version = select_compiler_versions(p, compilers)
-            append!(flags, String["-I/opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)",
-                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)/$(aatriplet(p))",
-                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)/backward",
-                                  "-I/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)/include",
-                                  "-I/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)/include-fixed",
-                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/include",
-                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root/include"])
+        if !Sys.isbsd(p) && !isnothing(gcc_version)
+            append!(flags, String["-isystem /opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)",
+                                  "-isystem /opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)/$(aatriplet(p))",
+                                  "-isystem /opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)/backward",
+                                  "-isystem /opt/$(aatriplet(p))/$(aatriplet(p))/include",
+                                  "-isystem /opt/$(aatriplet(p))/$(aatriplet(p))/sys-root/include"])
         end
 
         return flags
@@ -452,11 +453,12 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
                 # Use libgcc as the C runtime library
                 "-rtlib=libgcc"
             ])
-            gcc_version, llvm_version = select_compiler_versions(p, compilers)
-            append!(flags, String["-L/opt/$(aatriplet(p))/lib/gcc/opt/$(aatriplet(p))/lib/gcc",
-                                  "-L/opt/$(aatriplet(p))/$(aatriplet(p))/lib",
-                                  "-L/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)",
-                                  "-L/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root/lib",])
+            if !isnothing(gcc_version)
+                append!(flags, String["-L/opt/$(aatriplet(p))/lib/gcc/opt/$(aatriplet(p))/lib/gcc",
+                                    "-L/opt/$(aatriplet(p))/$(aatriplet(p))/lib",
+                                    "-L/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)",
+                                    "-L/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root/lib",])
+            end
         end
         # we want to use a particular linker with clang.  But we want to avoid warnings about unused
         # flags when just compiling, so we put it into "linker-only flags".
@@ -476,7 +478,6 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
     function macos_gcc_flags!(p::AbstractPlatform, flags::Vector{String} = String[])
         # On macOS, if we're on an old GCC, the default -syslibroot that gets
         # passed to the linker isn't calculated correctly, so we have to manually set it.
-        gcc_version, llvm_version = select_compiler_versions(p, compilers)
         if gcc_version.major in (4, 5)
             push!(flags, "-Wl,-syslibroot,/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root")
         end
@@ -496,11 +497,7 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
 
     function gcc_flags!(p::AbstractPlatform, flags::Vector{String} = String[])
         # Force proper cxx11 string ABI usage w00t w00t!
-        if cxxstring_abi(p) == "cxx11"
-            push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=1")
-        elseif cxxstring_abi(p) == "cxx03"
-            push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=0")
-        end
+        add_cxx_abi(p, flags)
 
         # Simulate some of the `__OSX_AVAILABLE()` macro usage that is broken in GCC
         if Sys.isapple(p) && something(os_version(p), v"14") < v"16"
@@ -536,7 +533,6 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
             append!(flags, get_march_flags(arch(p), march(p), "gcc"))
         end
         sanitize_compile_flags!(p, flags)
-        gcc_version, llvm_version = select_compiler_versions(p, compilers)
         if libc(platform) == "musl" && gcc_version in (v"4.8.5", v"5.2.0")
             add_system_includedir(flags)
         end
@@ -546,7 +542,6 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
     function gcc_link_flags!(p::AbstractPlatform, flags::Vector{String} = String[])
         # Yes, it does seem that the inclusion of `/lib64` on `powerpc64le` was fixed
         # in GCC 6, broken again in GCC 7, and then fixed again for GCC 8 and 9
-        gcc_version, llvm_version = select_compiler_versions(p, compilers)
         if arch(p) == "powerpc64le" && Sys.islinux(p) && gcc_version.major in (4, 5, 7)
             append!(flags, String[
                 "-L/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root/lib64",
@@ -1321,12 +1316,15 @@ function runner_setup!(workspaces, mappings, workspace_root, verbose, kwargs, pl
     rb = filter(s -> s.name == "RustBase", shards)
     rust_version = length(rb) == 1 ? only(rb).version : nothing
 
+    # Determine version of GCC toolchain.
+    gcc = filter(s -> s.name == "GCCBootstrap" && platforms_match(s.target, platform), shards)
+    gcc_version = length(gcc) == 1 ? only(gcc).version : nothing
     # Construct environment variables we'll use from here on out
     platform::Platform = get_concrete_platform(platform; compilers..., extract_kwargs(kwargs, (:preferred_gcc_version,:preferred_llvm_version))...)
     envs::Dict{String,String} = merge(platform_envs(platform, src_name; rust_version, verbose, compilers...), extra_env)
 
     # JIT out some compiler wrappers, add it to our mounts
-    generate_compiler_wrappers!(platform; bin_path=compiler_wrapper_path, compilers..., extract_kwargs(kwargs, (:allow_unsafe_flags,:lock_microarchitecture))...)
+    generate_compiler_wrappers!(platform; bin_path=compiler_wrapper_path, gcc_version, compilers..., extract_kwargs(kwargs, (:allow_unsafe_flags,:lock_microarchitecture))...)
     push!(workspaces, compiler_wrapper_path => "/opt/bin")
 
     if isempty(bootstrap_list)

--- a/src/Runner.jl
+++ b/src/Runner.jl
@@ -322,16 +322,17 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
             "--sysroot=/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root",
         ])
         if !Sys.isbsd(p)
-            append!(flags, [
-                # Find GCC toolchain here (for things like libgcc_s)
-                "--gcc-toolchain=/opt/$(aatriplet(p))"
-            ])
             if iscxx
                 append!(flags, [
                     # Link with libstdc++ when compiling c++ on non-BSDs
                     "-stdlib=libstdc++"
             ])
+            push!(flags, "-nobuiltininc")
             end
+        end
+        if Sys.islinux(p)
+            gcc_version, llvm_version = select_compiler_versions(p, compilers)
+            append!(flags, ["--gcc-install-dir=/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)"])
         end
         if Sys.iswindows(p)
             windows_cflags!(p, flags)
@@ -415,6 +416,18 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
         if Sys.isfreebsd(p)
             add_system_includedir(flags)
         end
+
+        if !Sys.isbsd(p)
+            gcc_version, llvm_version = select_compiler_versions(p, compilers)
+            append!(flags, String["-I/opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)",
+                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)/$(aatriplet(p))",
+                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)/backward",
+                                  "-I/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)/include",
+                                  "-I/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)/include-fixed",
+                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/include",
+                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root/include"])
+        end
+
         return flags
     end
 
@@ -434,6 +447,11 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
                 # Use libgcc as the C runtime library
                 "-rtlib=libgcc"
             ])
+            gcc_version, llvm_version = select_compiler_versions(p, compilers)
+            append!(flags, String["-L/opt/$(aatriplet(p))/lib/gcc/opt/$(aatriplet(p))/lib/gcc",
+                                  "-L/opt/$(aatriplet(p))/$(aatriplet(p))/lib",
+                                  "-L/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)",
+                                  "-L/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root/lib",])
         end
         # we want to use a particular linker with clang.  But we want to avoid warnings about unused
         # flags when just compiling, so we put it into "linker-only flags".

--- a/test/runners.jl
+++ b/test/runners.jl
@@ -199,7 +199,7 @@ end
             seekstart(iobuff)
             # Make sure `iobuff` contains only the input command, no other text
             is_broken = (compiler == "g++" && Sys.isapple(platform) && arch(platform) == "x86_64") # This gets fixed by using GCC7 or up
-            @test readchomp(iobuff) == string(cmd)
+            @test readchomp(iobuff) == string(cmd) broken=is_broken
         end
     end
 

--- a/test/runners.jl
+++ b/test/runners.jl
@@ -126,11 +126,10 @@ end
     # This tests only that compilers for all platforms can build and link simple C code
     @testset "Compilation - $(platform) - $(compiler)" for platform in platforms, compiler in ("cc", "gcc", "clang")
         mktempdir() do dir
-            # https://github.com/JuliaPackaging/BinaryBuilderBase.jl/issues/248
-            is_broken = compiler == "clang" && Sys.iswindows(platform)
             ur = preferred_runner()(dir; platform=platform)
             iobuff = IOBuffer()
             test_c = """
+                #include <stdlib.h>
                 int test(void) {
                     return 0;
                 }
@@ -155,18 +154,16 @@ end
                 $(compiler) -Werror -o main main.c -L. -ltest
                 """
             cmd = `/bin/bash -c "$(test_script)"`
-            @test run(ur, cmd, iobuff; tee_stream=devnull) broken=is_broken
+            @test run(ur, cmd, iobuff)
             seekstart(iobuff)
             # Make sure `iobuff` contains only the input command, no other text
-            @test readchomp(iobuff) == string(cmd) broken=is_broken
+            @test readchomp(iobuff) == string(cmd)
         end
     end
 
     # This tests only that compilers for all platforms can build and link simple C++ code
     @testset "Compilation - $(platform) - $(compiler)" for platform in platforms, compiler in ("c++", "g++", "clang++")
         mktempdir() do dir
-            # https://github.com/JuliaPackaging/BinaryBuilderBase.jl/issues/248
-            is_broken = compiler == "clang++" && Sys.iswindows(platform)
             ur = preferred_runner()(dir; platform=platform)
             iobuff = IOBuffer()
             needfpic = Sys.iswindows(platform) ? "" : "-fPIC"
@@ -198,11 +195,11 @@ end
                 $(compiler) -Werror -std=c++11 -o main main.cpp -L. -ltest
                 """
             cmd = `/bin/bash -c "$(test_script)"`
-            @test run(ur, cmd, iobuff; tee_stream=devnull) broken=is_broken
+            @test run(ur, cmd, iobuff; tee_stream=devnull)
             seekstart(iobuff)
             # Make sure `iobuff` contains only the input command, no other text
-            is_broken = is_broken || (compiler == "g++" && Sys.isapple(platform) && arch(platform) == "x86_64")
-            @test readchomp(iobuff) == string(cmd) broken=is_broken
+            is_broken = (compiler == "g++" && Sys.isapple(platform) && arch(platform) == "x86_64") # This gets fixed by using GCC7 or up
+            @test readchomp(iobuff) == string(cmd)
         end
     end
 

